src/main/java/com/google/devtools/build/lib/analysis/CommonPrerequisiteValidator.java

```diff
@@ -105,21 +108,46 @@ private void validateDirectPrerequisiteVisibility(
     if (!toolCheckAtDefinition
         || !attribute.isImplicit()
         || attribute.getName().equals(RuleClass.CONFIG_SETTING_DEPS_ATTRIBUTE)
-        || rule.getRuleClassObject().getRuleDefinitionEnvironmentLabel() == null) {
+        || !forStarlarkRuleOrAspect(context)) {
```

Comment:
Readability: rename `forStarlarkRuleOrAspect` to `inStarlarkRuleOrAspect` and move it into the RuleContext.Builder.